### DISCUSSION TITLE
What to do when a binary can't be identified?

### DISCUSSION THREAD

**Discussion by LoganDark** [redballoonsecurity/ofrak#477] [Q&A]
I have a firmware image I received directly from a manufacturer and when I drop it into OFRAK, all it says is basically "yep that definitely looks like a binary". The entire thing is just "[GenericBinary](https://ofrak.com/docs/reference/ofrak/core/binary.html#ofrak.core.binary.GenericBinary), [FilesystemEntry](https://ofrak.com/docs/reference/ofrak/core/filesystem.html#ofrak.core.filesystem.FilesystemEntry), [File](https://ofrak.com/docs/reference/ofrak/core/filesystem.html#ofrak.core.filesystem.File)".

When I try to unpack the programmer that consumes the firmware image, I get:

`Unable to find any analyzer for attributes ProgramAttributes
Filter: (ComponentTypeFilter(Analyzer) and ((AnalyzerOutputFilter(ProgramAttributes) and ((ComponentTypeFilter(Identifier) and ComponentTargetFilter(GhidraAutoLoadProject, GhidraProject, Pe, Program, GenericBinary, File, FilesystemEntry, LinkableBinary)) or (not ComponentTypeFilter(Identifier) and (ComponentTargetFilter(Pe) then ComponentTargetFilter(Program) then ComponentTargetFilter(GhidraAutoLoadProject, File, LinkableBinary) then ComponentTargetFilter(GhidraProject, GenericBinary, FilesystemEntry))))) then ((AnalyzerOutputFilter(ProgramAttributes) and ((ComponentTypeFilter(Identifier) and ComponentTargetFilter(GhidraAutoLoadProject, GhidraProject, Pe, Program, GenericBinary, File, FilesystemEntry, LinkableBinary)) or (not ComponentTypeFilter(Identifier) and (ComponentTargetFilter(Pe) then ComponentTargetFilter(Program) then ComponentTargetFilter(GhidraAutoLoadProject, File, LinkableBinary) then ComponentTargetFilter(GhidraProject, GenericBinary, FilesystemEntry))))))))`

What do I do now?

I'm using a Ghidra docker image compiled from the latest commit (d8e22a949782bdb48c24920138d8fda2962d9cc8).

**Comment by rbs-jacob** [comment]
Hey @LoganDark! It's possible that the OFRAK automatic identifiers cannot figure out what type of binary you have, but that doesn't mean that OFRAK cannot pull it apart. 

If you're using the GUI, the next step to try would be using the "Analyze" button, which will (among other things) try to run Binwalk on the firmware. The firmware attributes in the bottom left of the GUI will be amended to include whatever results Binwalk finds. These can be used for manually carving up the file into bits that OFRAK does know how to unpack. After that, you can export a script to automate the process and turn it into an OFRAK component. 

Do you already know anything about the firmware format?

**Reply by LoganDark** [reply]
Oh, I had always hit analyze first, thought that was implied, oops.

> Do you already know anything about the firmware format?

Unfortunately no. I don't know anything about the format or even the architecture that it's for. I don't even know if it uses some sort of encoding/encryption, and I can't use OFRAK to find out because it chokes when I try to feed it the programmer that consumes the firmware image. (which is a win32 binary)

**Reply by rbs-jacob** [reply]
Happy to work through reversing the firmware at a high level with you in this discussion if you're interested. 

Does Binwalk see anything it recognizes in the firmware?

The first thing that would help to judge whether it's encrypted or compressed is to send some images of the minimap on the right. The magnitude and entropy views will give a sense of what parts are encrypted or compressed. 

Probably there are issues with unpacking code regions from windows portable executable (PE/EXE) files in OFRAK, and that is why it is having trouble when you try to put the programmer in. If you just do "unpack" instead of "unpack recursively," how deep can you get before you see errors?

**Reply by LoganDark** [reply]
> Happy to work through reversing the firmware at a high level with you in this discussion if you're interested.

Sure!

> Does Binwalk see anything it recognizes in the firmware?

Running it from the command line, it gives me a completely empty table (headers, but no rows):

![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/b5e43f81-eea3-4105-96b8-e3e0983bffa9)

> The first thing that would help to judge whether it's encrypted or compressed is to send some images of the minimap on the right. The magnitude and entropy views will give a sense of what parts are encrypted or compressed.

![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/b64e0a7d-7cb5-477c-9ead-7efd953cefa2) ![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/557887c6-5edc-4402-accf-8c6944763bba)

It doesn't look encrypted but it still could be something dumb like an XOR.

> Probably there are issues with unpacking code regions from windows portable executable (PE/EXE) files in OFRAK, and that is why it is having trouble when you try to put the programmer in. If you just do "unpack" instead of "unpack recursively," how deep can you get before you see errors?

I never used unpack recursively. If I click analyze then unpack on the root entry (the executable itself) I get a list of things inside it, like the header sections and whatever, and one code section that looks like "code ( ... txt)". when I try to do either analyze or unpack on that (doesn't matter which) then at first it hangs for a few minutes, then pops up that error, and if I click analyze or unpack again the error pops up immediately with no wait.

**Reply by rbs-jacob** [reply]
To my eye, that doesn't look encrypted. A simple XOR is possible – could be one-byte (but I'd guess less than four-byte) repeating-key XOR (if any) given how much structure is retained. A good way to test this is to run `strings` (or the built-in OFRAK strings unpacker) to see if anything legible pops out.

Looks to me like code with not much padding. I also don't see anything that looks obviously like strings based on the magnitude view. Running `binwalk -A` will try to deduce the architecture. That could be a good next step.

**Reply by rbs-jacob** [reply]
For comparison, here are some ELF files in their minimap views.

<table>
<tr><td align="center">x86 ELF</td><td align="center">ARM ELF</td></tr>
<tr><td>

![image](https://github.com/redballoonsecurity/ofrak/assets/99368685/8871c5dd-da30-4a68-ad4b-0b319f0f9bbf)

</td><td>

![image](https://github.com/redballoonsecurity/ofrak/assets/99368685/70c5fece-99d7-460d-bf38-d1b8fce4d4a2)

</td></tr>
</table>

Photos taken from [this RBS blog post about the minimap views in OFRAK](https://redballoonsecurity.com/ofraks-minimap-view/).

**Reply by dannyp303** [reply]
Hi, judging by the byte maps you sent I would say this looks like code. `binwalk -A` shows function preludes and will detect which arch the binary is in. Additionally it seems strange to me that binwalk would pick absolutely nothing up for this, are there any identifying magic bytes at the begining, judging by the maps it looks like it starts with structured data.

**Reply by LoganDark** [reply]
> A good way to test this is to run `strings` (or the built-in OFRAK strings unpacker) to see if anything legible pops out.

![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/f9358426-d0a4-47d8-8e93-b18e2eece71d)

It looks almost like function name mangling?  at least at first...

![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/f3e26474-b18e-4781-9b34-4c0412231054)

> For comparison, here are some ELF files in their minimap views.

I don't think it'd be ELF considering this is probably straight machine code for a microcontroller.

> `binwalk -A` shows function preludes and will detect which arch the binary is in.

No dice unfortunately:

![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/eab0d761-96b1-478c-9de8-e5f360f68e3e)

> are there any identifying magic bytes at the begining, judging by the maps it looks like it starts with structured data.

The beginning bytes are: `30 07 00 20 15 b1 00 08 1d b1 00 08 1f b1 00 08`

I suppose it looks like machine code. Not sure if it's structured or not.

**Reply by dannyp303** [reply]
1) Those strings above are just peices of data that happen to be ascii printable, ignore those.
2) If binwalk doesn't find anything with -A I assume there is something funky going on with this binary since it definitely looks like structured code. 
3) Those `00 08` pieces look suspect. How far does that pattern continue, including the `00 20` in the begining. Expression for the pattern im speaking of being something like `<random byte> <random byte> 00 <(small number)>`

**Reply by LoganDark** [reply]
> 1\. Those strings above are just peices of data that happen to be ascii printable, ignore those.

I figured

> 3\. Those `00 08` pieces look suspect.

It's even `b1 00 08` all three times :3 

Hmmm...

![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/d22914e8-d172-46b7-aa82-203496757c51)

Bunch of ints, or machine code instructions? (Could `30 07` be a magic number, or is `30 07 00 20` an instruction?)

**Reply by LoganDark** [reply]
![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/4a98d47d-1efd-40af-bb3c-f64eba689045)
oh, here is more stuff that has `00 20` in it, forgot about that too. (just above but taller)

<details><summary>a</summary>

```
3007002015b100081db100081fb10008
21b1000823b1000825b1000800000000
00000000000000000000000027b10008
29b10008000000002bb100082db10008
2fb100082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
2fb100082fb10008a78600082fb10008
2fb100082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
154300082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
2fb10008b78700082fb100082fb10008
2fb100082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
2fb100082fb100082fb100082fb10008
dff810d007f018f900480047314f0008
aff300803007002010b505f0f5fb0121
072006f08bf8012144f20d6006f07df8
bde810404ff4804006f007b82de9f843
0324adf800400020dff818848df80300
6946404605f079fe4ff40070adf80000
8df802401c208df803006946fe4f3846
05f06bfe3e15adf8006004258df80350
6946384605f061fe4ff48450adf80000
8df803506946384605f057fe6802adf8
00008df803506946404605f04efeadf8
005010258df8035002208df802006946
384605f042fe0421384605f0aefe6003
adf800008df8024014208df803006946
404605f032feadf800608df803508df8
02406946404605f028fe3146404605f0
94fe0820adf800008df803508df80240
6946384605f019fe0821384605f085fe
8020adf800008df8024018208df80300
6946404605f009fe0121cc4805f094fe
bde8f88338b54ff48041480305f0eafb
4ff4a06005f0b2fb25208df800000025
```

</details>

and here is me trying to do anything with it


https://github.com/redballoonsecurity/ofrak/assets/4723091/e430aadc-aced-47bf-b55f-63ca8f9bbc8c

**Reply by dannyp303** [reply]
upon further inspection, this looks like an exception vector table. If this is true, this binary is a raw binary, where addr 0 is the literal start of the memory space. It could be thumb, which binwalk doesn't pick up with `-A`, but there are other ways. One of those ways being to just throw it into ghidra (maybe starting at addr 0x8000000) and see if it can decompile anything.

**Reply by LoganDark** [reply]
yeah this is probably flashed straight to a microcontroller. some part of it (maybe end of the memory space) probably contains a reset vector, but I looked there and I don't see anything that looks like an address

![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/4224de13-bd88-493f-a8c1-fb8809169391)

this is the end of the thing btw, it ends before even 0x8000

**Reply by LoganDark** [reply]
> One of those ways being to just throw it into ghidra

![image](https://github.com/redballoonsecurity/ofrak/assets/4723091/873f0ef3-7481-406f-b7e9-145692e195a5)

I wish I could throw it into ghidra but it makes me select an instruction set and I don't know which instruction set this is

**Reply by dannyp303** [reply]
When i say base address I mean all addresses would be `addr + <base address>` so it wouldn't need to be gigs of binary. Typically chips designate address spaces for different things and things starting with 8 are very common, all of this can be found in the corresponding chips user manual. 
[Example](https://www.nxp.com/docs/en/reference-manual/CFPRM.pdf)

That being said, the binary is very small and there is a possibility that this is for any of the dozens of niche micro-controller architectures that exist. Getting access to the user manual would help immensely. If you don't have a device, you can possibly find the FCC-ID and look up the publicly available pictures of the internals (if it transmits).
[This website is great](https://fcc.report/)

**Reply by LoganDark** [reply]
I do have the device, maybe I'll take it apart sometime and look at the chip inside. will do later
